Changes in module 'nf-core/mageck/test'
--- modules/nf-core/mageck/test/main.nf
+++ modules/nf-core/mageck/test/main.nf
@@ -1,5 +1,5 @@
 process MAGECK_TEST {
-    tag "$meta.id"
+    tag "${meta.treatment}_${meta.reference}"
     label 'process_medium'
 
     conda "bioconda::mageck=0.5.9"
@@ -21,14 +21,18 @@
 
     script:
     def args = task.ext.args ?: ''
+    def args2 = task.ext.args2 ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+
 
     """
     mageck  \\
         test \\
         $args \\
+        $args2 $meta.treatment \\
+        -c $meta.reference \\
         -k $count_table \\
-        -n $prefix
+        -n ${meta.treatment}_${meta.reference}
 
 
     cat <<-END_VERSIONS > versions.yml
@@ -36,4 +40,4 @@
         mageck: \$(mageck -v)
     END_VERSIONS
     """
-}
+}
************************************************************
