Changes in module 'nf-core/mageck/count'
--- modules/nf-core/mageck/count/main.nf
+++ modules/nf-core/mageck/count/main.nf
@@ -1,19 +1,22 @@
 process MAGECK_COUNT {
     tag "$meta.id"
-    label 'process_medium'
+    label 'process_high'
 
-    conda "${moduleDir}/environment.yml"
+
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mageck:0.5.9.5--py39h1f90b4d_3':
         'biocontainers/mageck:0.5.9.5--py39h1f90b4d_3' }"
 
     input:
-    tuple val(meta), path(inputfile)
+    tuple val(meta), path(fastq1), path(fastq2)
     path(library)
 
     output:
-    tuple val(meta), path("*count.txt")           , emit: count
+    tuple val(meta), path("*count.txt")            , emit: count
     tuple val(meta), path("*.count_normalized.txt"), emit: norm
+    tuple val(meta), path("*.countsummary.txt")    , emit: summary
+    tuple val(meta), path("*.count_normalized.txt"), emit: normalized
+    tuple val(meta), path("*.log")                 , emit: logs
     path "versions.yml"                            , emit: versions
 
     when:
@@ -22,8 +25,13 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def input_file = ("$inputfile".endsWith(".fastq.gz")) ? "--fastq ${inputfile}" : "-k ${inputfile}"
-    def sample_label = ("$inputfile".endsWith(".fastq.gz") || "$inputfile".endsWith(".fq.gz")) ? "--sample-label ${meta.id}" : ''
+    def sample_label = ("$fastq1".endsWith(".fastq.gz") || "$fastq1".endsWith(".fq.gz")) ? "--sample-label ${meta.id}" : ''
+    
+    if (meta.single_end && ("$fastq1".endsWith(".fastq.gz") || "$fastq1".endsWith(".fq.gz")) || "$fastq1".endsWith(".bam")) {
+        input = "--fastq $fastq1" 
+    } else {
+        input = "--fastq $fastq1 --fastq-2 $fastq2" 
+    }
 
     """
     mageck \\
@@ -32,7 +40,7 @@
         -l $library \\
         -n $prefix \\
         $sample_label \\
-        $input_file \\
+        $input \\
 
 
     cat <<-END_VERSIONS > versions.yml
@@ -40,17 +48,4 @@
         mageck: \$(mageck -v)
     END_VERSIONS
     """
-    stub:
-    def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
-    def input_file = ("$inputfile".endsWith(".fastq.gz")) ? "--fastq ${inputfile}" : "-k ${inputfile}"
-    def sample_label = ("$inputfile".endsWith(".fastq.gz") || "$inputfile".endsWith(".fq.gz")) ? "--sample-label ${meta.id}" : ''
-    """
-    touch ${prefix}.count.txt
-    touch ${prefix}.count_normalized.txt
-    cat <<-END_VERSIONS > versions.yml
-    "${task.process}":
-        mageck: \$(mageck -v)
-    END_VERSIONS
-    """
-}
+}
************************************************************
