Changes in module 'nf-core/medaka'
--- modules/nf-core/medaka/main.nf
+++ modules/nf-core/medaka/main.nf
@@ -11,8 +11,8 @@
     tuple val(meta), path(reads), path(assembly)
 
     output:
-    tuple val(meta), path("*.fa.gz"), emit: assembly
-    path "versions.yml"             , emit: versions
+    tuple val(meta), path("*_medakaConsensus.fasta"), emit: assembly
+    path "versions.yml"                             , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -21,16 +21,27 @@
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
+    if [[ "${reads.extension}" == "gz" ]]; then
+        gzip -df $reads
+        reads=$reads.baseName
+    else
+        reads=$reads
+    fi
+    if [[ "${assembly.extension}" == "gz" ]]; then
+        gzip -df $assembly
+        assembly=$assembly.baseName
+    else
+        assembly=$assembly
+    fi
+
     medaka_consensus \\
         -t $task.cpus \\
         $args \\
-        -i $reads \\
-        -d $assembly \\
+        -i \$reads \\
+        -d \$assembly \\
         -o ./
 
-    mv consensus.fasta ${prefix}.fa
-
-    gzip -n ${prefix}.fa
+    mv consensus.fasta ${prefix}.fasta
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
