Changes in module 'nf-core/mageck/mle'
--- modules/nf-core/mageck/mle/main.nf
+++ modules/nf-core/mageck/mle/main.nf
@@ -1,15 +1,14 @@
 process MAGECK_MLE {
-    tag "$meta.id"
-    label 'process_medium'
+    tag "$prefix"
+    label 'process_high'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/mageck:0.5.9--py37h6bb024c_0':
-        'biocontainers/mageck:0.5.9--py37h6bb024c_0' }"
+        'https://depot.galaxyproject.org/singularity/mageck:0.5.9.5--py39h1f90b4d_3':
+        'biocontainers/mageck:0.5.9.5--py39h1f90b4d_3' }"
 
     input:
-    tuple val(meta), path(count_table)
-    path(design_matrix)
+    tuple val(meta), path(design_matrix), path(count_table)
 
     output:
     tuple val(meta), path("*.gene_summary.txt") , emit: gene_summary
@@ -21,7 +20,8 @@
 
     script:
     def args = task.ext.args ?: ''
-    def prefix = task.ext.prefix ?: "${meta.id}"
+    prefix = meta.id ?: "${meta.treatment}_vs_${meta.reference}"
+    def design_command = design_matrix ? "-d $design_matrix" : ''
 
     """
     mageck \\
@@ -29,8 +29,9 @@
         $args \\
         --threads $task.cpus \\
         -k $count_table \\
-        -d $design_matrix \\
-        -n $prefix
+        -n $prefix     \\
+        $design_command 
+        
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
