Changes in module 'nf-core/mageck/mle'
--- modules/nf-core/mageck/mle/main.nf
+++ modules/nf-core/mageck/mle/main.nf
@@ -8,8 +8,7 @@
         'biocontainers/mageck:0.5.9--py37h6bb024c_0' }"
 
     input:
-    tuple val(meta), path(count_table)
-    path(design_matrix)
+    tuple val(meta), path(count_table), path(design_matrix)
 
     output:
     tuple val(meta), path("*.gene_summary.txt") , emit: gene_summary
@@ -27,7 +26,6 @@
     mageck \\
         mle \\
         $args \\
-        --threads $task.cpus \\
         -k $count_table \\
         -d $design_matrix \\
         -n $prefix

************************************************************
