Changes in module 'nf-core/minimap2/align'
--- modules/nf-core/minimap2/align/meta.yml
+++ modules/nf-core/minimap2/align/meta.yml
@@ -25,11 +25,6 @@
       description: |
         List of input FASTA or FASTQ files of size 1 and 2 for single-end
         and paired-end data, respectively.
-  - meta2:
-      type: map
-      description: |
-        Groovy Map containing reference information
-        e.g. [ id:'test_ref']
   - reference:
       type: file
       description: |

--- modules/nf-core/minimap2/align/main.nf
+++ modules/nf-core/minimap2/align/main.nf
@@ -9,8 +9,7 @@
         'biocontainers/mulled-v2-66534bcbb7031a148b13e2ad42583020b9cd25c4:1679e915ddb9d6b4abda91880c4b48857d471bd8-0' }"
 
     input:
-    tuple val(meta), path(reads)
-    tuple val(meta2), path(reference)
+    tuple val(meta), path(reads), path(reference)
     val bam_format
     val cigar_paf_format
     val cigar_bam

************************************************************
